src/assets/scripts/client/aircraft/FlightManagementSystem/Fms.js

```diff
@@ -0,0 +1,481 @@
+import LegModel from './LegModel';
+import { routeStringFormatHelper } from '../../navigationLibrary/Route/routeStringFormatHelper';
+import {
+    MCP_MODE,
+    MCP_MODE_NAME,
+    MCP_PROPERTY_MAP
+} from '../ModeControl/modeControlConstants';
+
+/**
+ *
+ *
+ * This class should always be instantiated from an `AircraftInstanceModel` and
+ * always instantiated with some form of a `spawnPatternModel`.
+ *
+ * @class Fms
+ */
+export default class Fms {
+    /**
+     * @constructor
+     * @param aircraftInitProps {object}
+     * @param initialRunwayAssignment {string}
+     * @param navigationLibrary {NavigationLibrary}
+     */
+    constructor(aircraftInitProps, initialRunwayAssignment, typeDefinitionModel, navigationLibrary) {
+        if (!_isObject(aircraftInitProps) || _isEmpty(aircraftInitProps)) {
+            throw new TypeError('Invalid aircraftInitProps passed to Fms');
+        }
+
+        /**
+         * Instance of the `NavigationLibrary`
+         *
+         * provides access to the aiport SIDs, STARs and Fixes via collection objects and fascade methods.
+         * used for route building
+         *
+         * @property _navigationLibrary
+         * @type {NavigationLibrary}
+         * @private
+         */
+        this._navigationLibrary = navigationLibrary;
+
+        /**
+         * Keeps track of current Altitude, Heading and Speed modes and values.
+         *
+         * @property _modeController
+         * @type {ModeController}
+         */
+        this._modeController = new ModeController(typeDefinitionModel);
+
+        /**
+         *
+         *
+         * @property _aircraftTypeDefinition
+         * @type {AircraftTypeDefinitionModel}
+         * @private
+         */
+        this._aircraftTypeDefinition = typeDefinitionModel;
+
+        /**
+         * Name of the initial runway assigned for takeoff/landing.
+         *
+         * This value is likely to change as an aircraft moves through the airspace.
+         *
+         * @property _runway
+         * @type {string}
+         * @private
+         */
+        this._runwayName = initialRunwayAssignment;
+
+        /**
+         * Collection of `LegModel` objects
+         *
+         * A `LegModel` represents a section of a flight plan:
+         * - single `WaypointModel` without restrictions (not part of a standard procedure)
+         * - single `WaypointModel` assigned to hold at
+         * - standard procedure (sid/star), which may contain may `WaypointModel` objects,
+         *   each which may contain restrictions
+         *
+         * @property legCollection
+         * @type {array}
+         * @default []
+         */
+        this.legCollection = [];
+
+        /**
+         * Current flight phase of an aircraft
+         *
+         * Currently only supports `arrival` and `departure`
+         *
+         * @property currentPhase
+         * @type {string}
+         * @default ''
+         */
+        this.currentPhase = '';
```

Comment:
This seems like something that may belong more to the aircraft than the fms. But if you don't agree, it's okay to stay here.